core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -864,20 +866,16 @@ object IO {
    */
   final def bracket0[E, A, B](
     acquire: IO[E, A]
-  )(release: (A, Option[Either[E, B]]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
-    Ref[Option[(A, Option[Either[E, B]])]](None).flatMap { m =>
-      (for {
-        a <- acquire
-              .flatMap(a => m.set(Some((a, None))).const(a))
-              .uninterruptibly
-        b <- use(a).attempt.flatMap(
-              eb =>
-                m.set(Some((a, Some(eb)))) *> (eb match {
-                  case Right(b) => IO.now(b)
-                  case Left(e)  => fail[E](e)
-                })
-            )
-      } yield b).ensuring(m.get.flatMap(_.fold(unit) { case (a, r) => release(a, r) }))
+  )(release: (A, ExitResult[E, B]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
+    Ref[Option[ExitResult[E, B]]](None).flatMap { m =>
+      for {
+        a <- acquire.uninterruptibly
+        b <- (for {
+              f <- use(a).fork
+              _ <- f.finished(r => m.set(Some(r)))
+              b <- f.join
+            } yield b).ensuring(m.get.flatMap(_.fold(unit)(release(a, _))))
+      } yield b
```

Comment:
This simplification is not correct (and it makes me want to add tests to ensure we don't break anything!). Because even though `acquire` is uninterruptible, the fiber can be interrupted before it starts the `b` (i.e. it can be interrupted between the `a <-` and `b <-` lines). Which will lead to the resource being acquired, but never released.

You need the old tricks of putting the ensuring on the outer level and making sure the ref is set right after the acquire, inside the uninterruptible block. It's more complex but it's the only correct way to write this.

Reply:
This just bit me in the a**, so very happy you caught it! I can definitely add a test for this in this PR.